src/k8s/pkg/client/kubernetes/certificate_signing_request.go

```diff
@@ -0,0 +1,79 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/canonical/k8s/pkg/log"
+	certificatesv1 "k8s.io/api/certificates/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// WatchCertificateSigningRequest watches a CertificateSigningRequest with the
+// given name and calls a verify function on each event.
+// WatchCertificateSigningRequest will continue watching the CSR until the
+// verify function returns true or an non-retriable error occurs.
+//
+// The verify function should return true if the CSR is valid and processing
+// should stop.
+// The verify function should return false if the CSR is not yet valid and
+// processing should continue.
+// The verify function should return an error if the CSR is in an invalid state
+// (e.g., failed or denied) or the issued certificate is invalid.
+func (c *Client) WatchCertificateSigningRequest(ctx context.Context, name string, verify func(csr *certificatesv1.CertificateSigningRequest) (bool, error)) error {
+	log := log.FromContext(ctx)
+	for {
+		w, err := c.CertificatesV1().CertificateSigningRequests().Watch(ctx, metav1.SingleObject(metav1.ObjectMeta{Name: name}))
+		if err != nil {
+			log.V(1).Info("Failed to watch CSR", "error", err)
+			continue
+		}
+
+		if retry, err := c.watchCertificateSigningRequestEvents(ctx, w, name, verify); err != nil {
+			return fmt.Errorf("failed to watch CSR %s: %w", name, err)
+		} else if !retry {
+			return nil
+		}
+
+		w.Stop()
+		log.V(1).Info("Retrying to watch CSR")
```

Comment:
I do not think this log line is useful in the grand scheme of things here. If we failed, we will print the error and retry. This line, without even mentioning the CSR name, is not useful and will be just noise.

Think: in a worker node where we have run the refresh-certs command, this will just print `Retrying to watch CSR` 3 times (since we have 3 CSRs) in the logs, without much context about which CSR, why, etc.

Remember that logs are an interface for someone _not_ familiar with the code to understand what is happening. Just printing messages like this will mostly lead to goose chases. Let's only log when we want to communicate something important to the end user.